stablehlo/transforms/StablehloRefineParameters.cpp

```diff
@@ -0,0 +1,226 @@
+      }
+
+      // Validate static dimension sizes match
+      for (auto [dimSize, refinedDimSize] :
+           llvm::zip(tensorType.getShape(), refinedTensorType.getShape())) {
+        if (!ShapedType::isDynamic(dimSize) && dimSize != refinedDimSize) {
+          return func.emitOpError(
+                     "refinement dimension sizes must match for static "
+                     "dimensions ")
+                 << type << " -> " << refinedType;
+        }
+      }
+    }
+  }
+  return success();
+}
+
+// Wrap operands in "type barriers" so the rest of the program remains valid
+// after the signature update and before shape refinement
+//
+// %1 = stablehlo.custom_call
+//   @stablehlo.shape_refinement_operand_wrapper(%arg1, %0)
+//     {indices_of_shape_operands = dense<1> : tensor<1xi64>}
+//     : (tensor<5x10xf32>, tensor<2xi64>) -> tensor<?x10xf32>
+//
+// When introduced this is an identity operation from:
+//   (tensor<?x10xf32>) -> tensor<?x10xf32>
+// But after the operand is changed, this op has type signature
+//  (tensor<5x10xf32>) -> tensor<?x10xf32>
+//
+// Before shape refinement, all future uses of this argument expect type
+// tensor<?x10xf32>. By updating these uses to instead use the wrapper, the IR
+// remains valid in the intermediate state.
+void wrapRefinedOperands(func::FuncOp func, TypeRange refinedTypes) {
+  Region& body = func.getBody();
+  OpBuilder builder(body);
+  builder.setInsertionPointToStart(&body.front());
+  for (int i = 0; i < static_cast<int>(body.getNumArguments()); ++i) {
+    BlockArgument arg = body.getArgument(i);
+    Type argType = arg.getType();
+    Type refinedType = refinedTypes[i];
+    if (argType != refinedType) {
+      auto customCall =
+          makeShapeRefinementOperandWrapper(builder, arg, refinedType);
+      auto callResult = customCall.getResult(0);
+      arg.replaceAllUsesExcept(callResult, customCall);
+    }
+  }
+}
+
+void refineOperandsAndUpdateFunctionSignature(func::FuncOp func,
+                                              TypeRange refinedTypes) {
+  Region& body = func.getBody();
+  OpBuilder builder(body);
+  for (int i = 0; i < static_cast<int>(body.getNumArguments()); ++i) {
+    auto arg = body.getArgument(i);
+    arg.setType(refinedTypes[i]);
+  }
+  func.setType(builder.getFunctionType(refinedTypes, func.getResultTypes()));
+}
+
+struct StablehloRefineParametersPass
+    : public impl::StablehloRefineParametersPassBase<
+          StablehloRefineParametersPass> {
+  using Super =
+      impl::StablehloRefineParametersPassBase<StablehloRefineParametersPass>;
+  StablehloRefineParametersPass() : Super() {}
+  StablehloRefineParametersPass(
+      const StablehloRefineParametersPassOptions& opts)
+      : Super(opts) {}
+  StablehloRefineParametersPass(TypeRange refinedTypes) : Super() {
+    this->refinedTypes = llvm::to_vector(refinedTypes);
+  }
+
+  void runOnOperation() override {
+    auto func = getStablehloRefineShapesTarget(getOperation());
```

Comment:
Maybe this should be renamed to `getMainOrUniqueFunction` or something since it's being used in more than one pass now. (I think the interpreter has similar logic, too.)

Reply:
Going to hold off on that rename for now as well. I.e. what if we wanted to provide the ability to refine an unused function, if there were a way to have multiple entrypoints.

It's a weak argument, but as I was making this change I saw the code comment for this function and looks like it was fairly deliberately named